[partitioner] New configuration for considering custom operations in knapsack optimization (#167083)

basilwong opened the pull request:
Summary:

Creates new configuration: `include_custom_and_triton_kernel_wrappers_in_knapsack_optimization`

 ---

If config toggle is set to true, then nodes meeting the following criteria no longer just recomputed by default:

```
op.namespace not in {"aten", "prim", "prims"}
```

or

```
node.target == torch.ops.higher_order.triton_kernel_wrapper_functional
```

Test Plan:
# E2E Testing

### Baseline

Here is the [tlparse link](https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/aps-omnifm_0908_6kx_fakecomm_base01-828d285622/attempt_0/version_0/rank_0/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000) for the [baseline training run](https://fburl.com/mlhub/bdmauhug). Observe that for [graph_11, it does not contain triton kernels](https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/aps-omnifm_0908_6kx_fakecomm_base01-828d285622/attempt_0/version_0/rank_0/-_11_0_0/min_cut_information_58.json?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000).

```
  "Recomputable Banned Nodes Order": [
    "addmm_2",
    "addmm_4",
    "addmm_7",
    "addmm_6",
    "addmm",
    "addmm_1",
    "addmm_3",
    "addmm_8",
    "addmm_9",
    "bucketize",
    "bucketize_3",
    "bucketize_9",
    "bucketize_24",
    "bucketize_27",
    "ge",
    "ge_1",
    "ge_2",
    "ge_3",
    "ge_4",
    "ge_5",
    "ge_6",
    "ge_7",
    "ge_8",
    "ge_9"
  ]
```

### Test: Config Set to False(Default)

Here is the [tlparse link](https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/aps-omnifm_0908_6kx_fakecomm_updated2-608c18f4ae/attempt_1/version_0/rank_0/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000&fbclid=IwY2xjawOdijtleHRuA2FlbQIxMQBicmlkETFoalJ3cG5lOTRFS2RDWElQc3J0YwZhcHBfaWQBMAABHkxt0N11P8qkVV5uz19tLQCqLdLrgLEBuZPqmIiqZ4XnTOn0ZR1JU-6-acPX_aem_leYEAAWYUXVq9xezW37bcA) for the [test training run](https://fburl.com/mlhub/u5ebucs3). Observe that for [graph_11, it does not contain triton kernels](https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/aps-omnifm_0908_6kx_fakecomm_updated2-608c18f4ae/attempt_1/version_0/rank_0/-_11_0_0/min_cut_information_58.json?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000&fbclid=IwY2xjawOdijtleHRuA2FlbQIxMQBicmlkETFoalJ3cG5lOTRFS2RDWElQc3J0YwZhcHBfaWQBMAABHkxt0N11P8qkVV5uz19tLQCqLdLrgLEBuZPqmIiqZ4XnTOn0ZR1JU-6-acPX_aem_leYEAAWYUXVq9xezW37bcA).

```
  "Recomputable Banned Nodes Order": [
    "addmm_2",
    "addmm_4",
    "addmm_7",
    "addmm_6",
    "addmm",
    "addmm_1",
    "addmm_3",
    "addmm_8",
    "addmm_9",
    "bucketize",
    "bucketize_3",
    "bucketize_9",
    "bucketize_24",
    "bucketize_27",
    "ge",
    "ge_1",
    "ge_2",
    "ge_3",
    "ge_4",
    "ge_5",
    "ge_6",
    "ge_7",
    "ge_8",
    "ge_9"
  ]
```

### Test: Config Set to True

Here is the [tlparse link](https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/aps-omnifm_0908_6kx_fakecomm_updated3-2d9a31068f/attempt_0/version_0/rank_0/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000) for the [test training run](https://fburl.com/mlhub/qlscz25s). Observe that for [graph_11, it does contain triton kernels](https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/aps-omnifm_0908_6kx_fakecomm_updated3-2d9a31068f/attempt_0/version_0/rank_0/-_11_0_0/min_cut_information_58.json?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000).

```
  "Recomputable Banned Nodes Order": [
    "triton_kernel_wrapper_functional_proxy_8",
    "triton_kernel_wrapper_functional_proxy_4",
    "addmm_2",
    "triton_kernel_wrapper_functional_proxy_11",
    "triton_kernel_wrapper_functional_proxy_13",
    "addmm_4",
    "triton_kernel_wrapper_functional_proxy_10",
    "addmm_7",
    "addmm_6",
    "triton_kernel_wrapper_functional_proxy_2",
    "triton_kernel_wrapper_functional_proxy",
    "triton_kernel_wrapper_functional_proxy_15",
    "offsets_range_2",
    "triton_kernel_wrapper_functional_proxy_17",
    "triton_kernel_wrapper_functional_proxy_6",
    "addmm",
    "addmm_1",
    "offsets_range_7",
    "addmm_3",
    "addmm_8",
    "addmm_9",
    "offsets_range_4",
    "offsets_range_5",
    "offsets_range_6",
    "asynchronous_complete_cumsum",
    "asynchronous_complete_cumsum_1",
    "asynchronous_complete_cumsum_2",
    "asynchronous_complete_cumsum_3",
    "asynchronous_complete_cumsum_4",
    "asynchronous_complete_cumsum_5",
    "asynchronous_complete_cumsum_6",
    "asynchronous_complete_cumsum_7",
    "asynchronous_complete_cumsum_8",
    "asynchronous_complete_cumsum_9",
    "bucketize",
    "offsets_range",
    "bucketize_3",
    "offsets_range_1",
    "bucketize_9",
    "offsets_range_3",
    "bucketize_24",
    "offsets_range_8",
    "bucketize_27",
    "offsets_range_9",
    "ge",
    "ge_1",
    "ge_2",
    "ge_3",
    "ge_4",
    "ge_5",
    "ge_6",
    "ge_7",
    "ge_8",
    "ge_9"
  ]
```

Differential Revision: D85683496

basilwong commented:
@pytorchbot label "topic: not user facing"

Lucaskabela commented on torch/_functorch/partitioners.py:
```diff
@@ -285,6 +286,19 @@ def _is_tangent(node: fx.Node) -> bool:
     return node.op == "placeholder" and "tangents" in str(node.target)
 
 
+def _include_custom_ops(node: fx.Node) -> bool:
```
I think we are saying include any non-builtin node here, so maybe instead of `_include_custom_ops`, a name of `include_non_builtin_nodes` better captures these semantics, and we can include a block comment here saying the main use case for this is custom ops/triton wrappers

basilwong replied:
Sounds good, I'll make the changes so it's more clear.

Lucaskabela reviewed (changes requested):
Some thoughts on naming, but the core logic looks good :) Feel free to re-request once addressed

Lucaskabela reviewed (changes requested):
See comments - just some more minor naming

basilwong committed:
[partitioner] New configuration for considering custom operations in knapsack optimization (#169636)

Summary:


Creates new configuration: `is_non_builtin_to_include`

---

If config toggle is set to true, then nodes meeting the following criteria no longer just recomputed by default:

```
op.namespace not in {"aten", "prim", "prims"}
```

or

```
node.target == torch.ops.higher_order.triton_kernel_wrapper_functional
```

Test Plan:
# E2E Testing


### Baseline

Here is the [tlparse link](https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/aps-omnifm_0908_6kx_fakecomm_base01-828d285622/attempt_0/version_0/rank_0/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000) for the [baseline training run](https://fburl.com/mlhub/bdmauhug). Observe that for [graph_11, it does not contain triton kernels](https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/aps-omnifm_0908_6kx_fakecomm_base01-828d285622/attempt_0/version_0/rank_0/-_11_0_0/min_cut_information_58.json?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000).

```
  "Recomputable Banned Nodes Order": [
    "addmm_2",
    "addmm_4",
    "addmm_7",
    "addmm_6",
    "addmm",
    "addmm_1",
    "addmm_3",
    "addmm_8",
    "addmm_9",
    "bucketize",
    "bucketize_3",
    "bucketize_9",
    "bucketize_24",
    "bucketize_27",
    "ge",
    "ge_1",
    "ge_2",
    "ge_3",
    "ge_4",
    "ge_5",
    "ge_6",
    "ge_7",
    "ge_8",
    "ge_9"
  ]
```

### Test: Config Set to False(Default)

Here is the [tlparse link](https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/aps-omnifm_0908_6kx_fakecomm_updated2-608c18f4ae/attempt_1/version_0/rank_0/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000&fbclid=IwY2xjawOdijtleHRuA2FlbQIxMQBicmlkETFoalJ3cG5lOTRFS2RDWElQc3J0YwZhcHBfaWQBMAABHkxt0N11P8qkVV5uz19tLQCqLdLrgLEBuZPqmIiqZ4XnTOn0ZR1JU-6-acPX_aem_leYEAAWYUXVq9xezW37bcA) for the [test training run](https://fburl.com/mlhub/u5ebucs3). Observe that for [graph_11, it does not contain triton kernels](https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/aps-omnifm_0908_6kx_fakecomm_updated2-608c18f4ae/attempt_1/version_0/rank_0/-_11_0_0/min_cut_information_58.json?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000&fbclid=IwY2xjawOdijtleHRuA2FlbQIxMQBicmlkETFoalJ3cG5lOTRFS2RDWElQc3J0YwZhcHBfaWQBMAABHkxt0N11P8qkVV5uz19tLQCqLdLrgLEBuZPqmIiqZ4XnTOn0ZR1JU-6-acPX_aem_leYEAAWYUXVq9xezW37bcA).

```
  "Recomputable Banned Nodes Order": [
    "addmm_2",
    "addmm_4",
    "addmm_7",
    "addmm_6",
    "addmm",
    "addmm_1",
    "addmm_3",
    "addmm_8",
    "addmm_9",
    "bucketize",
    "bucketize_3",
    "bucketize_9",
    "bucketize_24",
    "bucketize_27",
    "ge",
    "ge_1",
    "ge_2",
    "ge_3",
    "ge_4",
    "ge_5",
    "ge_6",
    "ge_7",
    "ge_8",
    "ge_9"
  ]
```

### Test: Config Set to True

Here is the [tlparse link](https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/aps-omnifm_0908_6kx_fakecomm_updated3-2d9a31068f/attempt_0/version_0/rank_0/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000) for the [test training run](https://fburl.com/mlhub/qlscz25s). Observe that for [graph_11, it does contain triton kernels](https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/aps-omnifm_0908_6kx_fakecomm_updated3-2d9a31068f/attempt_0/version_0/rank_0/-_11_0_0/min_cut_information_58.json?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000).

```
  "Recomputable Banned Nodes Order": [
    "triton_kernel_wrapper_functional_proxy_8",
    "triton_kernel_wrapper_functional_proxy_4",
    "addmm_2",
    "triton_kernel_wrapper_functional_proxy_11",
    "triton_kernel_wrapper_functional_proxy_13",
    "addmm_4",
    "triton_kernel_wrapper_functional_proxy_10",
    "addmm_7",
    "addmm_6",
    "triton_kernel_wrapper_functional_proxy_2",
    "triton_kernel_wrapper_functional_proxy",
    "triton_kernel_wrapper_functional_proxy_15",
    "offsets_range_2",
    "triton_kernel_wrapper_functional_proxy_17",
    "triton_kernel_wrapper_functional_proxy_6",
    "addmm",
    "addmm_1",
    "offsets_range_7",
    "addmm_3",
    "addmm_8",
    "addmm_9",
    "offsets_range_4",
    "offsets_range_5",
    "offsets_range_6",
    "asynchronous_complete_cumsum",
    "asynchronous_complete_cumsum_1",
    "asynchronous_complete_cumsum_2",
    "asynchronous_complete_cumsum_3",
    "asynchronous_complete_cumsum_4",
    "asynchronous_complete_cumsum_5",
    "asynchronous_complete_cumsum_6",
    "asynchronous_complete_cumsum_7",
    "asynchronous_complete_cumsum_8",
    "asynchronous_complete_cumsum_9",
    "bucketize",
    "offsets_range",
    "bucketize_3",
    "offsets_range_1",
    "bucketize_9",
    "offsets_range_3",
    "bucketize_24",
    "offsets_range_8",
    "bucketize_27",
    "offsets_range_9",
    "ge",
    "ge_1",
    "ge_2",
    "ge_3",
    "ge_4",
    "ge_5",
    "ge_6",
    "ge_7",
    "ge_8",
    "ge_9"
  ]
```

Differential Revision: D85683496
